src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -411,13 +412,42 @@ private static object CreateInstance([DynamicallyAccessedMembers(DynamicallyAcce
                 throw new InvalidOperationException(SR.Format(SR.Error_CannotActivateAbstractOrInterface, type));
             }
 
-            if (!type.IsValueType)
+            ConstructorInfo[] constructors = type.GetConstructors(DeclaredOnlyLookup);
+
+            bool hasParameterlessPublicConstructor =
+                constructors.Any(ctor => ctor.IsPublic && ctor.GetParameters().Length == 0);
+
+            if (constructors.Length > 0 && !hasParameterlessPublicConstructor)
             {
-                bool hasDefaultConstructor = type.GetConstructors(DeclaredOnlyLookup).Any(ctor => ctor.IsPublic && ctor.GetParameters().Length == 0);
-                if (!hasDefaultConstructor)
+                // if the only constructor is private, then throw
+                if (constructors.Length == 1 && !constructors[0].IsPublic)
                 {
                     throw new InvalidOperationException(SR.Format(SR.Error_MissingParameterlessConstructor, type));
                 }
+
+                // find the biggest constructor so that we can bind to the most parameters
+                ParameterInfo[] parameters = constructors[0].GetParameters();
+
+                int indexOfChosenConstructor = 0;
+
+                for (int index = 1; index < constructors.Length; index++)
+                {
+                    ParameterInfo[] constructorParameters = constructors[index].GetParameters();
+                    if (constructorParameters.Length > parameters.Length)
```

Comment:
I feel like we shouldn't rely on constructor order to break ties. Perhaps that could result in an exception?

Reply:
Good point. If no suitable constructor is found, it falls back to `Activator.CreateInstance`, which will throw 'no parameterless constructor'